// angularfire2/providers/firebase_list.js
var core_1 = require('angular2/core');
var angularfire2_1 = require('../angularfire2');
var absolute_path_resolver_1 = require('../utils/absolute_path_resolver');
var firebase_list_factory_1 = require('../utils/firebase_list_factory');
function FirebaseList(config) {
    var normalConfig = normalizeConfig(config);
    return new core_1.Provider(normalConfig.token, {
        useFactory: function (defaultFirebase) { return firebase_list_factory_1.FirebaseListFactory(absolute_path_resolver_1.absolutePathResolver(defaultFirebase, normalConfig.path)); },
        deps: [angularfire2_1.FirebaseUrl]
    });
}
exports.FirebaseList = FirebaseList;
function normalizeConfig(config) {
    switch (typeof config) {
        case 'string':
            var strConf = config;
            return {
                token: strConf,
                path: strConf
            };
        case 'object':
            var conf = config;
            var token;
            if (conf.token) {
                token = conf.token;
            }
            else if (conf.path) {
                token = conf.path;
            }
            else {
                token = FirebaseList;
            }
            return {
                token: token,
                path: conf.path
            };
        default:
            return {
                token: FirebaseList,
                path: ''
            };
    }
}
exports.normalizeConfig = normalizeConfig;
